### ISSUE TITLE
Memoize with currying

### ISSUE THREAD

**Issue by youldali** [ramda/ramda#2633]
Hi,

I was wondering if there's was a way, or if it's planned in the future that the memoize function somehow supports a curried function, returning the final result instead of the intermediate function.

For example: const fA = curry( (argA, argB) => 'curried')
If we apply memoization on it: memoizeWith (R.identity, fA)
And we call fA supplying only the 1st argument: const appliedfA = fA('1st arg')
then a cache key is created returning the intermediate function
=> all subsequent calls to appliedfA('2ndArg') are not memoized

Thanks

**Comment by CrossEye** [comment]
No, there is no way to do so now and there is no plan to do so.  I can't see how it would work: how would we know when the user chose to cache a function versus when they want second-level calls to be cached?  Any suggestions are welcome.  But I don't think there is a huge appetite for working too much with memoization.

**Comment by youldali** [comment]
I guess a memoized function that curries the function passed, and caches the final result could do the trick, maybe like so:

```
const memoizeWithCurrying= (keyGenerationFunction: Function, f: Function) => {
    const cache = {};
    return curryN(
        f.length,
        (...args) => {
            const key = keyGenerationFunction(...args);
            return cache[key] ? cache[key] : (cache[key] = f(...args) );
        }
    );
};
```

**Comment by CrossEye** [comment]
That looks suspiciously like `memoizeWith`.

**Comment by mirandaleandro** [comment]
Hi, folks. I'm new to Ramda. 

This question is making me wonder what is the rationale around key generation. In the curry example, I'm thinking there is no need since each function has arity of 1, and maybe there is a case when arity is >1. Can anyone share the reasoning with me?

Thank you and **congratulations on building such a nice lib & community around it**

**Comment by CrossEye** [comment]
Key generation is necessary since whether we cache with an `Object` or with a `Map`, we need primitive keys.  For the former, we can really only use strings/symbols.  For the latter, any primitive will do.  (While `Map`s allow arbitrary keys, they depend on reference equality, and that will rarely work well with Ramda.)  But people could be passing arbitrary data to their functions; we we need to be able to turn that into a key.  Ramda would have no idea how to do that for arbitrary data.  Hence, we leave it to the user.

> Thank you and congratulations on building such a nice lib & community around it

Thanks.  It's nice to be appreciated!
